### PR TITLE
Add ability to limit purge query results

### PR DESCRIPTION
**Pull request by exell-christopher**
-Add configurable limits to purge queries
-Add indexes to be used by purge queries
-Move to use of Concurrent HashSet for session id store instead of synchronizing around it so that scavenge doesn't block session creation/removal.

https://bugs.eclipse.org/bugs/show_bug.cgi?id=464839

### CODE REVIEW

**Comment by joakime** [comment]
Pull request pending signed CLA at the Eclipse Foundation.

File: jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java [hunk 1]
```diff
@@ -489,7 +525,7 @@ public boolean idInUse(String sessionId)
         /*
          * optimize this query to only return the valid variable
          */
-        DBObject o = _sessions.findOne(new BasicDBObject("id",sessionId), __valid_true);
+        DBObject o = _sessions.findOne(new BasicDBObject("id",sessionId));
```
Comment (janbartel) [inline jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java]: Why did you remove the limitation of this query to just pulling back the "valid" field instead of the whole session?

File: jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java [hunk 2]
```diff
@@ -106,15 +107,19 @@
...
-    protected final Set<String> _sessionsIds = new HashSet<String>();
-    
-    
+    protected final Set<String> _sessionsIds = new ConcurrentHashSet<>();
+
+
+    /**
+     * The maximum number of items to return from a purge query.
+     */
+    private int __purgeLimit = 0;
```
Comment (janbartel) [inline jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java]: A minor thing: double underscore prefix is for statics only.

File: jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java [hunk 3]
```diff
@@ -179,55 +184,60 @@ public MongoSessionIdManager(Server server, DBCollection sessions)
                 BasicDBObjectBuilder.start().add("id",1).get(),
                 BasicDBObjectBuilder.start().add("unique",true).add("sparse",false).get());
         _sessions.ensureIndex(
-                BasicDBObjectBuilder.start().add("id",1).add("version",1).get(),
-                BasicDBObjectBuilder.start().add("unique",true).add("sparse",false).get());
+                BasicDBObjectBuilder.start().add("id", 1).add("version", 1).get(),
```
Comment (janbartel) [inline jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java]: There's quite a few places in the patch where it seems that just whitespaces changes have been made, making it a little difficult to see if any actual code change has been made.

File: jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java (hunk 1 above)
Comment (exell-christopher) [inline jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java]: I can back that out. Accidental removal when re-orging after adding the indexes. Oddly that doesn't seem to overly impact perf.

File: jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java (hunk 3 above)
Comment (exell-christopher) [inline jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java]: Sorry about that, apparently IntelliJ 14 changed the way it strips white space. Didn't notice it until I put in the pull request. I can try and back those out if you like.

File: tests/test-sessions/test-mongodb-sessions/src/test/java/org/eclipse/jetty/nosql/mongodb/PurgeValidSessionTest.java [hunk 4]
```diff
@@ -123,6 +123,75 @@ public void testPurgeValidSession() throws Exception
     }
 
 
+    @Test
+    public void testPurgeValidSessionWithPurgeLimitSet() throws Exception
```
Comment (janbartel) [inline tests/test-sessions/test-mongodb-sessions/src/test/java/org/eclipse/jetty/nosql/mongodb/PurgeValidSessionTest.java]: Ran this test locally and it failed at line 175: there were more sessions in the db than expected. Sure you've got the config right here for the purgeFully() to work on line 154?

Comment (exell-christopher) [inline tests/test-sessions/test-mongodb-sessions/src/test/java/org/eclipse/jetty/nosql/mongodb/PurgeValidSessionTest.java]: I'll take another look. This test runs fine locally for me.

Comment (janbartel) [inline tests/test-sessions/test-mongodb-sessions/src/test/java/org/eclipse/jetty/nosql/mongodb/PurgeValidSessionTest.java]: Worth noting that I ran that test as part of the whole suite of tests, not by itself. Maybe some of the other tests add sessions that don't expire.

Comment (exell-christopher) [inline tests/test-sessions/test-mongodb-sessions/src/test/java/org/eclipse/jetty/nosql/mongodb/PurgeValidSessionTest.java]: This should be fixed now. Added handling to deal with existing sessions in the store that aren't going to be invalidated.

File: jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java (hunk 3 above)
Comment (exell-christopher) [inline jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java]: Spacing cleaned up

File: jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java (hunk 2 above)
Comment (exell-christopher) [inline jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java]: Fixed.

File: jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java (hunk 1 above)
Comment (exell-christopher) [inline jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionIdManager.java]: Fixed.

**Comment by janbartel** [comment]
Changes were merged to master in commit 8e2011d
